### PR TITLE
compressed sync [DEVINFRA-491]

### PR DESCRIPTION
**Pull request by silverjam**
+ Add compressed (upload) sync support, files are compressed and replaced during sync so that resumption of an interrupted sync doesn't have to repeat work
  - (NB: if we didn't replace files we wouldn't be able to tell between an interrupted compression and a completed compression without some additional store of metadata)
+ Create new modules: `upload`, `sync` and `compression` -- could probably make a `download_decompressed` module to clean-up the feature flag mess in the `download` module
+ Add stages to ensure that the cli and lib can correctly build (and function) without features enabled
+ Fix-up compilation, linting and testing for the above changes

### CODE REVIEW

File: src/esthri/src/compression.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            let size = compressed.path().metadata()?.len();
+            debug!("new file size: {}", size);
+            Ok((compressed, size))
+        })
+        .await
+        .expect(EXPECT_SPAWN_BLOCKING)
+    }
+}
+
+pub async fn compress_and_replace(path: impl AsRef<Path>) -> Result<PathBuf> {
```
Comment (notoriaga) [inline src/esthri/src/compression.rs:60]: should this be `pub(crate)`?

Comment (silverjam) [reply src/esthri/src/compression.rs:60]: I exposed this as a subcommand in the esthri tool, it was mostly useful for helping me generate test data for the unit test, but can probably remove that subcommand now

Comment (silverjam) [reply src/esthri/src/compression.rs]: fixed

**Review by notoriaga — APPROVED** [review]

File: src/esthri/src/compression.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    let path = path.as_ref().to_path_buf();
+    async move {
+        task::spawn_blocking(move || {
+            let size = path.metadata()?.len();
+            debug!("old file size: {}", size);
+            debug!("compressing: {}", path.display());
+            let mut reader = GzEncoder::new(BufReader::new(file), Compression::default());
+            let mut compressed = NamedTempFile::new()?;
+            std::io::copy(&mut reader, &mut compressed)?;
+            compressed.flush()?;
```
[the remark this answers is not in the document]
Comment (silverjam) [reply src/esthri/src/compression.rs:49]: > also recently discovered this neat shorthand for `compressed.seek(SeekFrom::Start(0))?` - `compressed.rewind()?`

cool, but it's an unstable API for now: https://github.com/rust-lang/rust/issues/85149

[the remark this answers is not in the document]
Comment (notoriaga) [reply src/esthri/src/compression.rs:49]: 🤦‍♂️

File: src/esthri/src/ops/sync.rs [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+{
+    use std::str::FromStr;
+    input_stream.map(move |params| async move {
+        let (path, metadata) = params?;
+        let path = PathBuf::from_str(&path)?;
+        #[cfg(feature = "compression")]
+        {
+            use crate::compression::compress_and_replace;
+            let (path, local_etag) = if compressed {
+                if path.to_string_lossy().ends_with(".gz") {
```
[the remark this answers is not in the document]
Comment (silverjam) [reply src/esthri/src/ops/sync.rs:229]: fixed

File: src/esthri/src/compression.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            compressed.seek(SeekFrom::Start(0))?;
+            let size = compressed.path().metadata()?.len();
```
[the remark this answers is not in the document]
Comment (silverjam) [reply src/esthri/src/compression.rs:51]: fixed

File: src/esthri/src/compression.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (silverjam) [reply src/esthri/src/compression.rs:49]: > I don't think you need to explicitly call flush if you use `io::copy`

fixed

**Pull request merged**